tensorflow/lite/experimental/micro/examples/micro_speech/arduino-esp32/audio_provider.cc

```diff
@@ -0,0 +1,204 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/lite/experimental/micro/examples/micro_speech/audio_provider.h"
+
+#include <Arduino.h>        // NOLINT
+#include <driver/i2s.h>
+#include <soc/i2s_reg.h> // I2S_TIMING_REG
+
+#include "tensorflow/lite/experimental/micro/examples/micro_speech/micro_features/micro_model_settings.h"
+
+namespace {
+static const i2s_port_t i2s_num = I2S_NUM_0; // i2s port number
+
+// These are the raw buffers that are filled by the ADC during DMA
+constexpr int kAdcNumSlots = 8;
+constexpr int kAdcSamplesPerSlot = 128;
+
+TaskHandle_t xCaptureTaskHandle = NULL;
+
+bool g_is_audio_initialized = false;
+constexpr int kAudioCaptureBufferSize = kAudioSampleFrequency * 0.5;
+int32_t* g_audio_capture_buffer = (int32_t*)heap_caps_malloc(kAudioCaptureBufferSize * sizeof(int32_t), MALLOC_CAP_DMA | MALLOC_CAP_32BIT);
```

Comment:
should `g_audio_capture_buffer` and `g_audio_output_buffer` be initialized
inside a function when it is used for the first time?

Reply:
i still need some hint, what to do with the memory allocation (see discussion above). however i agree, this should be somewhere else and then there should be a check if it succeeded.